modules/cuml/src/umap.cpp

```diff
@@ -0,0 +1,283 @@
+  // env);
+  this->params_ = update_params(args[0]);
+  ML::UMAP::find_ab(handle, &this->params_);
+  this->embeddings_ = DeviceBuffer::wrapper_t();
+}
+
+void UMAP::fit(float* X,
+               cudf::size_type n_samples,
+               cudf::size_type n_features,
+               float* y,
+               int64_t* knn_indices,
+               float* knn_dists,
+               bool convert_dtype,
+               float* embeddings) {
+  raft::handle_t handle;
+  CUDA_CHECK(cudaStreamSynchronize(handle.get_stream()));
+
+  ML::UMAP::fit(
+    handle, X, y, n_samples, n_features, knn_indices, knn_dists, &this->params_, embeddings);
+}
+
+void UMAP::transform(float* X,
+                     cudf::size_type n_samples,
+                     cudf::size_type n_features,
+                     int64_t* knn_indices,
+                     float* knn_dists,
+                     float* orig_X,
+                     int orig_n,
+                     bool convert_dtype,
+                     float* embeddings,
+                     float* transformed) {
+  raft::handle_t handle;
+  CUDA_CHECK(cudaStreamSynchronize(handle.get_stream()));
+
+  ML::UMAP::transform(handle,
+                      X,
+                      n_samples,
+                      n_features,
+                      knn_indices,
+                      knn_dists,
+                      orig_X,
+                      orig_n,
+                      embeddings,
+                      n_samples,
+                      &this->params_,
+                      transformed);
+}
+
+Napi::Value UMAP::fit(Napi::CallbackInfo const& info) {
+  CallbackArgs args{info};
+
+  auto X           = DeviceBuffer::wrapper_t(*DeviceBuffer::Unwrap(args[0]));
+  auto y           = DeviceBuffer::IsInstance(args[3])
+                       ? reinterpret_cast<float*>(DeviceBuffer::Unwrap(args[3])->data())
+                       : nullptr;
+  auto knn_indices = DeviceBuffer::IsInstance(args[4])
+                       ? reinterpret_cast<int64_t*>(DeviceBuffer::Unwrap(args[4])->data())
+                       : nullptr;
+  auto knn_dists   = DeviceBuffer::IsInstance(args[5])
+                       ? reinterpret_cast<float*>(DeviceBuffer::Unwrap(args[5])->data())
+                       : nullptr;
+
+  auto embeddings = DeviceBuffer::wrapper_t(*DeviceBuffer::Unwrap(args[7]));
+
+  fit(reinterpret_cast<float*>(X->data()),
+      args[1],
+      args[2],
+      y,
+      knn_indices,
+      knn_dists,
+      args[6],
+      reinterpret_cast<float*>(embeddings->data()));
+
+  return embeddings;
+}
+
+Napi::Value UMAP::transform(Napi::CallbackInfo const& info) {
+  CallbackArgs args{info};
+
+  auto X           = DeviceBuffer::wrapper_t(*DeviceBuffer::Unwrap(args[0]));
+  auto knn_indices = DeviceBuffer::IsInstance(args[3])
+                       ? reinterpret_cast<int64_t*>(DeviceBuffer::Unwrap(args[3])->data())
+                       : nullptr;
```

Comment:
To simplify, we can make the variable always a DeviceBuffer wrapper, then use the conversion operator in the call below.

```suggestion
  DeviceBuffer:wrapper_t knn_indices = DeviceBuffer::IsInstance(args[3]) ? args[3] : DeviceBuffer::New(args.Env());
```